app/src/main/java/org/mozilla/fenix/home/recentvisits/RecentVisitsFeature.kt

```diff
@@ -0,0 +1,131 @@
+                    MAX_RESULTS_OF_EACH_TYPE
+                ).map {
+                    RecentHistoryHighlight(
+                        title = if (it.title.isNullOrBlank()) it.url else it.title!!,
+                        url = it.url
+                    )
+                }
+            }
+
+            val recentHistoryGroups = async {
+                historyMetadataStorage.getHistoryMetadataSince(Long.MIN_VALUE)
+                    .filter { it.totalViewTime > 0 && it.key.searchTerm != null }
+                    .groupBy { it.key.searchTerm!! }
+                    .mapValues { group ->
+                        // Within a group, we dedupe entries based on their url so we don't display
+                        // a page multiple times in the same group, and we sum up the total view time
+                        // of deduped entries while making sure to keep the latest updatedAt value.
+                        val metadataInGroup = group.value
+                        val metadataUrlGroups = metadataInGroup.groupBy { metadata -> metadata.key.url }
+                        metadataUrlGroups.map { metadata ->
+                            metadata.value.reduce { acc, elem ->
+                                acc.copy(
+                                    totalViewTime = acc.totalViewTime + elem.totalViewTime,
+                                    updatedAt = max(acc.updatedAt, elem.updatedAt)
+                                )
+                            }
+                        }
+                    }
+                    .map { (title, data) ->
+                        RecentHistoryGroup(
+                            title = title,
+                            historyMetadata = data
+                        )
+                    }
+                    .sortedByDescending { it.lastUpdated() }
+                    .take(MAX_RESULTS_OF_EACH_TYPE)
+            }
+
+            updateState(historyHighlights.await(), recentHistoryGroups.await())
+        }
+    }
+
+    @VisibleForTesting
+    internal fun updateState(
+        historyHighlights: List<RecentHistoryHighlight>,
+        historyGroups: List<RecentHistoryGroup>
+    ) {
+        homeStore.dispatch(
+            HomeFragmentAction.RecentHistoryChange(
+                getCombinedHistory(historyHighlights, historyGroups)
+            )
+        )
+    }
+
+    @VisibleForTesting
+    @Suppress("MagicNumber")
+    internal fun getCombinedHistory(
+        historyHighlights: List<RecentHistoryHighlight>,
+        historyGroups: List<RecentHistoryGroup>,
+    ): List<RecentlyVisitedItem> {
+        // Temporary algorithm while scoring is added to metadata history
+        // after which both types will be sortable by their score.
+        val recentGroups = historyGroups.take(3)
+        val recentHighlights = historyHighlights
+            .take((MAX_RESULTS_TOTAL - recentGroups.size).coerceAtMost(MAX_RESULTS_OF_EACH_TYPE))
+
+        return (recentGroups + recentHighlights)
```

Comment:
Just a first pass. There are two more cases we need to cover here:

- If there are not enough highlights we should "fill up" with groups. So 3 groups is the minimum, but we should still have 9 groups if there are no highlights, or 6 groups if there are 3 highlights.

- If a page is already in a displayed group, we should not also add it separately as a highlight. Otherwise, we can be in a situation where the highlights are (mostly) the same as the items in a group e.g. after extensive research on a topic.

Reply:
Everything else looks good to me, but I will take a closer look and test more!

Reply:
> If there are not enough highlights we should "fill up" with groups. So 3 groups is the minimum, but we should still have 9 groups if there are no highlights, or 6 groups if there are 3 highlights.

This differs a bit from the specs on Jira based on which I understood there's a hard limit of 6 items of a type.
(The hardcoded max 3 here was just a temporary approach until scoring is available for groups also).
Without that limit of 6 the algorithm then becomes `show a total of 9 groups and highlights distributed evenly or as many as available from each type`.
Since we have to show an uneven number of items of two types one type will always appear more than the other. Will be going with groups. Changing to favor highlights is easy.

One question I have related to groups: should we filter out groups only having only 1 metadata entry similar to [what we do with search groups for "Jump back in"](https://github.com/mozilla-mobile/fenix/blob/116a2f0c2bd9a279476ee0828a9f120e8870980b/app/src/main/java/org/mozilla/fenix/ext/BrowserState.kt#L118) ?

Reply:
Yes, it differs a bit because the edge case wasn't thought about that there could be no highlights at all, which isn't very likely. I mostly pointed it out because we should fill up with groups then. If you want to fill up just to get to 6 that's good enough too. We will likely always have highlights anyway and end up with the 3 most recent groups. (Until we introduce scoring for groups.)

> One question I have related to groups: should we filter out groups only having only 1 metadata entry similar to.
No, for history we decided that groups of 1 item are fine, but it's not a high confidence decision, just a separate debate, plus we'd have to change this in history as well. So, let's keep this independent. :)